pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -196,6 +196,11 @@ func (r *ReconcileNSTemplateSet) ensureNamespaceResource(logger logr.Logger, nsT
 			return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
 				"failed to set controller reference for namespace type '%s'", tcNamespace.Type)
 		}
+		ns := &corev1.Namespace{}
+		if err := r.scheme.Convert(obj, ns, nil); err != nil {
```

Comment:
@xcoulon thanks for finding this.  
Can you plz explain what was the issue and how this will help to fix it ??
Also, have you seen such sort of `Convert()` used to achieve similar thing elsewhere ??

Reply:
@nurali-techie when converting the `unstructured.Unstructured` object into a `corev1.Namespace`, the "fake client" gets this type of object to be created and stores it in its underlying `tracker`. When calling the `client.List(...)` method, the tracker is queried and it filters the objects with the expected  `APIVersion` and Kind`. However, the client's impl assumes that the object matches the following cases:
```
if src.Type().AssignableTo(dest.Type()) {
	dest.Set(src)
} else if src.Type().ConvertibleTo(dest.Type()) {
	dest.Set(src.Convert(dest.Type()))
} else {
	return fmt.Errorf("item[%d]: can't assign or convert %v into %v", i, src.Type(), dest.Type())
}

```
(see https://github.com/kubernetes/apimachinery/blob/ed135c5b96450fd24e5e981c708114fbbd950697/pkg/api/meta/help.go#L195-L201)

which is not the case with an `unstructured.Unstructured` object. 
Converting to a valid `corev1.Namespace` works for the fakeclient. I'm not sure we need such a workaround for the real client/clusters, though.